### PR TITLE
feat: add support for MasonryFlashlist

### PR DESCRIPTION
**Pull request by rafpato**
## Description

This PR adds support for the Masonry Flashlist Layout

## Motivation and Context

Since FlashList is already supported, this additional layout is a nice win (I also need it in my app). 
The API is pretty much the same except some props that are **not** supported:

- [horizontal](https://shopify.github.io/flash-list/docs/usage#horizontal)
- [inverted](https://shopify.github.io/flash-list/docs/usage#inverted)
- [initialScrollIndex](https://shopify.github.io/flash-list/docs/usage#initialscrollindex)
- [viewabilityConfigCallbackPairs](https://shopify.github.io/flash-list/docs/usage#viewabilityconfigcallbackpairs)
- [onBlankArea](https://shopify.github.io/flash-list/docs/usage#onblankarea)

## How Has This Been Tested?

I ran the example app on iOS and Android emulators, please find some videos below

| Android | iOS |
|--------|--------|
| <video src="https://github.com/user-attachments/assets/f05aa2a2-eb5b-48f7-9caf-e9997fcec20d" /> | <video src="https://github.com/user-attachments/assets/eacfd703-479a-447a-9d1b-7bcb65b68a71" />| 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have followed the guidelines in the README.md file.
- [x] I have updated the documentation as necessary.
- [x] My changes generate no new warnings.

## Additional Notes (Optional)

I wasn't sure what to do with the docs so I just duplicated the FlashList page for now but I think we should add there:
- which version this will be released in?
- perhaps update the compatibility matrix?

### CODE REVIEW

**Comment by rafpato** [comment]
hey @e-younan, made this a draft PR as I wanted to get some feedback from you first 🙏🏼

**Comment by e-younan** [comment]
Hey @rafpato, thanks for opening this! I'll review it now. 😄

File: example/src/screens/usage/MasonryFlashList.tsx [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  const { bottom } = useSafeAreaInsets();
+  const ref = useRef<MasonryFlashListRef<number>>(null);
+
+  const data = useMemo(() => range({ end: 500 }), []);
+
+  const renderItem: ListRenderItem<number> = useCallback(({ item }) => {
+    return (
+      <View style={styles.item}>
+        <Text style={styles.itemText}>{item}. Scroll to see header animation</Text>
+      </View>
```
Comment (e-younan) [inline example/src/screens/usage/MasonryFlashList.tsx:78]: nit: would be nice to render some randomly-sized images here 😄

**Review by e-younan — CHANGES_REQUESTED** [review]
Minor issues to fix, but looks good to me. On the `Getting Started` page of the documentation, you should extend the following section:

```
Here is another If you intend to use the [FlashListWithHeaders](http://localhost:3000/docs/components/flash-list-with-headers) component, please ensure that you review the [Compatibilty table](http://localhost:3000/docs/getting-started#compatibility) above and install the correct versions of each library.
```

to mention the new component you added.

File: src/components/containers/MasonryFlashList.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    <View
+      style={[
+        styles.container,
+        containerStyle,
+        !ignoreLeftSafeArea && { paddingLeft: insets.left },
+        !ignoreRightSafeArea && { paddingRight: insets.right },
+      ]}
+    >
+      {!absoluteHeader && HeaderComponent({ showNavBar, scrollY })}
+      <AnimatedFlashList
```
[the remark this answers is not in the document]
Comment (rafpato) [reply src/components/containers/MasonryFlashList.tsx]: a7de98794fc4030bd091789e95969504509d7caa

File: src/components/containers/MasonryFlashList.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+import { StyleSheet, View } from 'react-native';
+import { useSafeAreaInsets } from 'react-native-safe-area-context';
+import Animated, { useAnimatedRef } from 'react-native-reanimated';
+import { MasonryFlashList, MasonryFlashListProps, MasonryFlashListRef } from '@shopify/flash-list';
+
+import type { SharedScrollContainerProps } from '.';
+import FadingView from './FadingView';
+import { useScrollContainerLogic } from './useScrollContainerLogic';
+
+type AnimatedFlashListType<ItemT> = React.ComponentProps<
```
[the remark this answers is not in the document]
Comment (rafpato) [reply src/components/containers/MasonryFlashList.tsx]: a7de98794fc4030bd091789e95969504509d7caa

File: src/components/containers/MasonryFlashList.tsx [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  React.ComponentClass<Animated.AnimateProps<MasonryFlashListProps<ItemT>>, any>
+> &
+  SharedScrollContainerProps;
+
+const AnimatedFlashList = Animated.createAnimatedComponent(
```
[the remark this answers is not in the document]
Comment (rafpato) [reply src/components/containers/MasonryFlashList.tsx]: a7de98794fc4030bd091789e95969504509d7caa

File: example/src/screens/usage/MasonryFlashList.tsx (hunk 1 above)
Comment (rafpato) [reply example/src/screens/usage/MasonryFlashList.tsx]: f125a302e7156a27e0b15d82a59ab4777d3fcc8e

**Comment by rafpato** [comment]
> Minor issues to fix, but looks good to me. On the `Getting Started` page of the documentation, you should extend the following section:
> 
> ```
> Here is another If you intend to use the [FlashListWithHeaders](http://localhost:3000/docs/components/flash-list-with-headers) component, please ensure that you review the [Compatibilty table](http://localhost:3000/docs/getting-started#compatibility) above and install the correct versions of each library.
> ```
> 
> to mention the new component you added.

Thanks so much for the feedback! I've made all the requested changes 😇 

As for the documentation, I added that the component is only available in 0.14.0 assuming it would be available in the next version.

**Review by e-younan — APPROVED** [review]
Great work! 😄

**Pull request merged**